### DISCUSSION TITLE
Can eglot connect to a standalone Eclipse JDT language server running outside of Emacs?

### DISCUSSION THREAD

**Discussion by hxiong** [joaotavora/eglot#799] [Q&A] [answered]
Hi all,

New to Eglot. I have downloaded a standalone executable Eclipse JDT jar, and I'm able to use Eglot with Eclipse JDT Language Server if I execute the JDT LS from within Emacs Eglot. However, my question is, If I run the Eclipse JDT server manually in the terminal, how do I configure Eglot to connect to it and use it?

I tried to run the Eclipse JDT independently using:
`
java  \
-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=1044 \
-DCLIENT_PORT=8679
-Declipse.application=org.eclipse.jdt.ls.core.id1 \
-Dosgi.bundles.defaultStartLevel=4 \
-Declipse.product=org.eclipse.jdt.ls.core.product \
-Dlog.level=ALL \
-noverify \
-Xmx1G \
-jar ./plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar \
-configuration ./config_mac \
-data ./eclipse-jdt-data \
--add-modules=ALL-SYSTEM \
--add-opens java.base/java.util=ALL-UNNAMED \
--add-opens java.base/java.lang=ALL-UNNAMED \
`

The Eclipse JDT runs without any issue.

Then I tried to set Eglot to connect to it:
`(add-to-list 'eglot-server-programs '((java-mode) .
									  ("localhost" 8679)))
`
But this set up does not work. Does Eglot support such set up at the moment?

Cheers,
Ted

**Comment by skangas** [comment, marked as answer]
Could you explain what you mean by "doesn't work"? Do you get a backtrace if you try that after saying `M-x toggle-debug-on-error`?

**Reply by hxiong** [reply]
Good suggestion. Here is the trace I have:
`Debugger entered--Lisp error: (file-error "make client process failed" "Connection refused" :name "EGLOT (java-test/java-mode)" :buffer nil :host "localhost" :service 8679 :nowait nil :tls-parameters nil :coding nil)
  make-network-process(:name "EGLOT (java-test/java-mode)" :buffer nil :host "localhost" :service 8679 :nowait nil :tls-parameters nil :coding nil)
  open-network-stream("EGLOT (java-test/java-mode)" nil "localhost" 8679)
  apply(open-network-stream "EGLOT (java-test/java-mode)" nil "localhost" 8679 nil)
  #f(compiled-function () #<bytecode -0xd9db083314f36df>)()
  #f(compiled-function (cl--cnm conn slots) #<bytecode -0x1c46f989646e8fb2>)(#f(compiled-function (&rest cnm-args) #<bytecode 0x12388d69bf933f2a>) #<eglot-lsp-server eglot-lsp-server-1fe90dedf114> (:name "EGLOT (java-test/java-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91209ea2578c>) :request-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91e0e1e2578c>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0xd9db083314f36df>)))
  apply(#f(compiled-function (cl--cnm conn slots) #<bytecode -0x1c46f989646e8fb2>) #f(compiled-function (&rest cnm-args) #<bytecode 0x12388d69bf933f2a>) (#<eglot-lsp-server eglot-lsp-server-1fe90dedf114> (:name "EGLOT (java-test/java-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91209ea2578c>) :request-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91e0e1e2578c>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0xd9db083314f36df>))))
  #f(compiled-function (&rest args) #<bytecode 0x194eab558ec9a857>)(#<eglot-lsp-server eglot-lsp-server-1fe90dedf114> (:name "EGLOT (java-test/java-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91209ea2578c>) :request-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91e0e1e2578c>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0xd9db083314f36df>)))
  apply(#f(compiled-function (&rest args) #<bytecode 0x194eab558ec9a857>) #<eglot-lsp-server eglot-lsp-server-1fe90dedf114> (:name "EGLOT (java-test/java-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91209ea2578c>) :request-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91e0e1e2578c>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0xd9db083314f36df>)))
  initialize-instance(#<eglot-lsp-server eglot-lsp-server-1fe90dedf114> (:name "EGLOT (java-test/java-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91209ea2578c>) :request-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91e0e1e2578c>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0xd9db083314f36df>)))
  #f(compiled-function (class &rest slots) "Default constructor for CLASS `eieio-default-superclass'.\nSLOTS are the initialization slots used by `initialize-instance'.\nThis static method is called when an object is constructed.\nIt allocates the vector used to represent an EIEIO object, and then\ncalls `initialize-instance' on that object." #<bytecode -0x2a3f03cc2d92fca>)(eglot-lsp-server :name "EGLOT (java-test/java-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91209ea2578c>) :request-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91e0e1e2578c>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0xd9db083314f36df>))
  apply(#f(compiled-function (class &rest slots) "Default constructor for CLASS `eieio-default-superclass'.\nSLOTS are the initialization slots used by `initialize-instance'.\nThis static method is called when an object is constructed.\nIt allocates the vector used to represent an EIEIO object, and then\ncalls `initialize-instance' on that object." #<bytecode -0x2a3f03cc2d92fca>) eglot-lsp-server (:name "EGLOT (java-test/java-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91209ea2578c>) :request-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91e0e1e2578c>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0xd9db083314f36df>)))
  make-instance(eglot-lsp-server :name "EGLOT (java-test/java-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91209ea2578c>) :request-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91e0e1e2578c>) :on-shutdown eglot--on-shutdown :process #f(compiled-function () #<bytecode -0xd9db083314f36df>))
  apply(make-instance eglot-lsp-server :name "EGLOT (java-test/java-mode)" :events-buffer-scrollback-size 2000000 :notification-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91209ea2578c>) :request-dispatcher #f(compiled-function (server method params) #<bytecode 0xfee91e0e1e2578c>) :on-shutdown eglot--on-shutdown (:process #f(compiled-function () #<bytecode -0xd9db083314f36df>)))
  eglot--connect(java-mode (vc . "~/dc-ted/quick-x-starter/java-test/") eglot-lsp-server ("localhost" 8679) "java")
  eglot(java-mode (vc . "~/dc-ted/quick-x-starter/java-test/") eglot-lsp-server ("localhost" 8679) "java" t)
  funcall-interactively(eglot java-mode (vc . "~/dc-ted/quick-x-starter/java-test/") eglot-lsp-server ("localhost" 8679) "java" t)
  call-interactively(eglot record nil)
  command-execute(eglot record)
  execute-extended-command(nil "eglot")
`

**Reply by hxiong** [reply]
I think something is wrong with my set up. I read this blog post about how JDT Language Server works: https://ncona.com/2021/12/implementing-a-language-server-protocol-client. 

It sounds like to use JDT LS as a running instance, we need to start our LSP client (Eglot) and bind it to a port address (1234 for example). Then start the standalone JDT instance with the client's port address(-DCLIENT_PORT=1234). JDT will then initiate talking or handshake with the LSP client at the specified address. 

So it's JDT that starts the communication with the LSP client, not the other way around? With my set up, my LSP client (Eglot) tries to start the communication process with JDT server, so that won't work right?

Thanks.

**Reply by skangas** [reply]
eglot definitely needs to connect to the server. This sounds like a server issue, as the backtrace shows eglot trying and failing to connect to `localhost:8679`.

I'd try verifying with something like `nc localhost 8679` or `netstat --listen|grep 8679` if the server is listening to that port and that you can connect to it.
